commit c3dabd9cd8fdb29333b2a2b00560a4654d6665fb
Author: Alexander K <ak@natsys-lab.com>
Date:   Sun Jan 17 16:54:05 2016 +0300

    Proper initialization and update of socket CPU binding

diff --git a/include/net/sock.h b/include/net/sock.h
index 5c44cb0d..60a43920 100644
--- a/include/net/sock.h
+++ b/include/net/sock.h
@@ -849,16 +849,14 @@ static inline int sk_backlog_rcv(struct sock *sk, struct sk_buff *skb)
 	return sk->sk_backlog_rcv(sk, skb);
 }
 
+#define TFW_SK_CPU_INIT	USHRT_MAX
+
 static inline void sk_incoming_cpu_update(struct sock *sk)
 {
 #ifdef CONFIG_SECURITY_TEMPESTA
-	if (unlikely(sk->sk_incoming_cpu != raw_smp_processor_id()))
-		net_warn_ratelimited("Warning: bad socket cpu locality:"
-				     " old_cpu=%d curr_cpu=%d\n",
-				     sk->sk_incoming_cpu,
-				     raw_smp_processor_id());
+	if (sk->sk_incoming_cpu == TFW_SK_CPU_INIT)
 #endif
-	sk->sk_incoming_cpu = raw_smp_processor_id();
+		sk->sk_incoming_cpu = raw_smp_processor_id();
 }
 
 static inline void sock_rps_record_flow_hash(__u32 hash)
diff --git a/security/tempesta/Kconfig b/security/tempesta/Kconfig
index ba1859bc..0fa2cb46 100644
--- a/security/tempesta/Kconfig
+++ b/security/tempesta/Kconfig
@@ -2,6 +2,7 @@ config SECURITY_TEMPESTA
 	bool "Tempesta FW Support"
 	depends on SECURITY && NET && INET
 	select SECURITY_NETWORK
+	select RPS
 	select CRYPTO
 	select CRYPTO_HMAC
 	select CRYPTO_SHA1
