commit 6bc35c81baa28cb7aa7cf29af7afc5e7b8848f41
Author: Alexander K <ak@natsys-lab.com>
Date:   Wed Jul 11 01:25:09 2018 +0300

    Remove debugging message

diff --git a/net/ipv4/tcp_output.c b/net/ipv4/tcp_output.c
index 0eecab7c..01bd67d8 100644
--- a/net/ipv4/tcp_output.c
+++ b/net/ipv4/tcp_output.c
@@ -2372,7 +2372,6 @@ static bool tcp_write_xmit(struct sock *sk, unsigned int mss_now, int nonagle,
 			if (unlikely(sk->sk_write_xmit(sk, skb, limit)))
 				break;
 #endif
-		pr_err("AK_DBG %s: skb=%pK len=%u\n", __func__, skb, skb->len);
 		if (unlikely(tcp_transmit_skb(sk, skb, 1, gfp)))
 			break;
 
